### ISSUE TITLE
Analysis Server jump to definition on "Function" in a function type should take you to the Function class

### ISSUE THREAD

**Issue by natebosch** [dart-lang/sdk#50019]
Currently, placing the cursor on the word "Function" in `void Function()` and invoking jump to definition in the IDE results in no definition found. Jumping to the `class Function` definition in the SDK would be more useful than an error.

We should also expand the doc comment on Function to include an explanation of the function type syntax. cc @mit @lrhn What do you think?

**Comment by eernstg** [comment]
I tend to think that it would be slightly misleading to jump from `Function` in `void Function(int)` to the declaration of the class named `Function`: There is no reasonable interpretation of the function type `void Function(int)` that makes this word a reference to that class. You could say that it's similar to this situation:

```dart
var get = 2;

class C {
  int get g => 'Hello';
}
```

The word `get` occurs twice, and one of them is a declaration, but the other one does not refer to that declaration.

**Comment by bwilkerson** [comment]
I agree that the word `Function` is not, in this position, a reference to the class.

I also agree that it's unfortunate that the IDE reports that there is no declaration when on the word `Function`, but I'll note that it has the same behavior when on the keyword `class` in `class C {}`. I believe that the server is doing what the IDE expects in both cases, and is just trying to tell the user that there is no definition to go to so that they don't think go-to-definition is broken. I just think there might be a better way to do that.

I'm going to close this as WAI.

**Comment by natebosch** [comment]
I disagree that it is misleading. The word "Function" in `void Function()` _is_ related to the `class Function`, because it is defining a subtype.

The type  `Object?` is similarly subtype of the `class Object`, but when we jump to definition on "Object" within `Object?` we are taken to `class Object`.

**Comment by bwilkerson** [comment]
If it were true that the "Function" in `void Function()` is a reference to the class because it defines a subtype of `Function`, then it seems to me that would be a fairly strong argument for using "Record" in a record type to indicate that it's creating subtype of `Record`.

**Comment by natebosch** [comment]
> it seems to me that would be a fairly strong argument for using "Record" in a record type to indicate that it's creating subtype of `Record`.

That proposal is exactly why I filed this issue 😄 

If we do use the `Record(...)` syntax for record types, I'd also ask for jump to definition to work there, and I do think it's a pro for that option since may improve discoverability. We are currently leaning away from this syntax towards using whitespace to disambiguate annotations, but this has been discussed.

**Comment by bwilkerson** [comment]
> That proposal is exactly why I filed this issue

In that case, I'll re-open it.

> ... using whitespace to disambiguate annotations ...

😢

**Comment by lrhn** [comment]
If we could link from the `Function` in `int Function()` to the "class" `Function`, it wouldn't be *bad*.
It could provide useful documentation about functions in general, and the `... Function(...)` syntactic form in particular.

It might not be the definition of the letters `Function` in that position, but it's the most relevant declaration for a function type.

**Comment by natebosch** [comment]
https://dart-review.googlesource.com/c/sdk/+/260440 would expand the doc comment on `Function` to include a description of function type.
